First pass at feature parity for url add-on based on app.routes.

imalberto opened the pull request:
CHANGELOG:
- dispatcher has been modified to annotate the middleware with mojito specific attributes
- `route-maker.find()` and `route-maker.make()` methods have been refactored to work with `app.routes`
- `route-maker.getComputedRoutes()` provides a temporary BC for the mojito-client runtime to work, until the new client router lands
- applications can now register route like: `app.get('/foo', mojito.dispatch('bar.foo'));`,
- `Y.mojito.util.copy()` now support regex and function types

TODO:
- [x] fix unit tests
- [x] fix functional tests
- [x] fix examples

imalberto committed:
First pass at feature parity for url add-on based on app.routes.

imalberto committed:
Fix verb to be lowercase

caridy commented on lib/dispatcher.js:
```diff
@@ -56,6 +63,29 @@ module.exports = {
+                    routes = req.app.routes[method];
+                    if (routes && routes.length > 0) {
+                        routes.forEach(function (route) {
+                            cache.push({
```
this should only happens if annotations exists, `extractAnnotations(route)`, other routes are not relevant for mojito.

caridy commented on lib/dispatcher.js:
```diff
@@ -47,7 +48,13 @@ module.exports = {
 
-        return function (req, res, next) {
+        function extractAnnotations(route) {
+            var fn = route && route.callbacks && route.callbacks.length &&
```
in the future we might want to support a more generic annotation mechanism, walking all `callbacks` and producing an array of annotations, for now looking at the last one is just fine.

caridy commented:
need more time to review the addon, too complex.

imalberto committed:
`ac.url.make()` now appends query string to the url for parameters not part of the parametrized `route.path`

imalberto committed:
Updates after first review pass.

imalberto committed:
Less expensive to calculate requires keys if there is a matched route vs doing it for all routes in the constructor.

caridy commented on lib/app/addons/ac/deploy.server.js:
```diff
@@ -116,7 +117,13 @@ YUI.add('mojito-deploy-addon', function (Y, NAME) {
             }
 
-            clientConfig.routes = store.getRoutes(contextClient);
+            // NOTE:
```
what is temporary here?

caridy commented on lib/app/addons/ac/deploy.server.js:
```diff
@@ -116,7 +117,13 @@ YUI.add('mojito-deploy-addon', function (Y, NAME) {
+            // This is temporary until the new client router lands.
+            // clientConfig.routes = store.getRoutes(contextClient);
+            //
+            // TODO: pageRoutes probably should be sanitized before shipping to
```
sanitized how? whatever we need to do, we should do it upfront as a one time operation, so when it comes to serialize and send, we have everything ready.

caridy commented on lib/dispatcher.js:
```diff
@@ -56,6 +66,33 @@ module.exports = {
                 return next(new Error('app.mojito does not exist.'));
             }
 
+            if (mojito.routes && Object.keys(mojito.routes).length === 0) {
```
this condition could be always true if there are no routes. I propose to have mojito.routes as undefined, then once we build the cache, we set it to an object, and that's the condition.

caridy commented on lib/dispatcher.js:
```diff
@@ -56,6 +66,33 @@ module.exports = {
             }
 
+            if (mojito.routes && Object.keys(mojito.routes).length === 0) {
+                debug('building up route cache for first time');
```
remove " for first time".

caridy commented on lib/dispatcher.js:
```diff
@@ -115,7 +160,8 @@ module.exports = {
         outputHandler.page.appConfig = store.getAppConfig(context);
         // routes are not contextualized anymore
-        outputHandler.page.routes = store.getRoutes();
+        // outputHandler.page.routes = store.getRoutes();
```
just remove it

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,253 +6,268 @@
+            });
+
+            // handle left over params by appending them to query string
             for (k in params) {
```
there is a bug here, probably an existing bug. If I have two keys: `foo` and `foobar`, and you try to replace `:foo` first, then it will invalidate `:foobar` as well. The solution is to sort the lookup array so you try to replace the longest keys first.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,253 +6,268 @@
-
-            var parts = query.split('?'),
+            var my = this,
+                parts = query.split('?'),
```
are the existing querystrings irrelevant? I don't see the second part of `parts` being used.

imalberto replied:
Thanks, added the check here https://github.com/yahoo/mojito/pull/1191/files#L1R269 . Querystring as part of the query will take priority.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -267,173 +282,51 @@ YUI.add('mojito-route-maker', function(Y, NAME) {
-            }
-
-            return match;
+            return route;
```
This route is, I presume, the new route object from cache, which is probably different from what they get today. Can we transform it?

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,253 +6,268 @@
+    @param {Object} appRoutes routes object from express
+    @return {Object} 
+    **/
+    function convertRoutes(appRoutes) {
```
Can you sync up with Julien to see if what are they doing with the routes? If they are not using anything fancy, we can remove this.

imalberto committed:
Updated based on PR#1191 feedback.

imalberto committed:
Updates from PR#1191 feedback + unit tests

caridy commented:
Ok, let's take this for a spin.

+1

imalberto committed:
- Fixed Y.mojito.util.copy to deal with regexp and function types
- Updated apidocs
- Added unit tests

imalberto committed:
Fix functional tests configuration for:
- lazyloadclientmojitclient
- lazyloadrpc

imalberto committed:
Fixed functional tests test/func/routing

imalberto committed:
Fixed 3 more fund tests in tests/func/routing

imalberto committed:
- Fixed the glob descriptors in run.js when using the --descriptor argument
- Fixed the routing related functional tests
- Fixed examples/developerguide

imalberto committed:
Misc bug fixes discovered while working on the functional tests:

- ac.config.getRoutes() now return route configuration in the same format as ac.url.find()
- the "name" property of a route is now constructed by concatenating the "path" + "call" together, to ensure that the "name" is unique for those route entries that share the same mojit instance and action.

imalberto commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,254 +6,299 @@
+
+        verbs[verb] = true;
+
+        name = route.path + '.' + (route.dispatch && route.dispatch.call);
```
Prepending the "path" to the "call", so that this can be supported without having name conflict:

```
// In this case, "/foo" and "/bar" are sharing the "admin.foo" controller
app.get('/foo', mojito.dispatch('admin.foo'));
app.get('/bar', mojito.dispatch('admin.foo'));
```

caridy replied:
Is the name used to hash this into a collection? at this method, name is just another member, I don't see why it has to be unique.

caridy replied:
I see `convertAppRoutes` is hashing it using name. I don't think this is good because internally you will never reference that by a similar name. It should be an array, just like it is in express.

imalberto replied:
I think the original thought of setting the "name" to just be "call" is good enough.

Making the name unique is more of a "maintain backward compatibility" as previously, the "name" was provided by developer in `routes.json`, and by convention, they were unique across all routes.

Turns out that mojito-shaker uses the name property to group rollups per route entry. We can revert it to being set to "call" only, and document that in order to ensure uniqueness, the caller should generate and associate a named identifier to it.

imalberto commented on lib/app/autoload/util.common.js:
```diff
@@ -169,13 +169,20 @@ YUI.add('mojito-util', function(Y, NAME) {
+                type,
                 copy = Y.mojito.util.copy;
 
             if (!oldObj || typeof oldObj !== 'object') {
```
Avoiding an invocation of `Y.Lang.type()` if it is not required.

caridy commented on examples/developer-guide/configure_routing/mojits/RoutingMojit/controller.server.js:
```diff
@@ -14,21 +14,34 @@ YUI.add('RoutingMojit', function(Y, NAME) {
-            routes.root_route.verbs.forEach(function(n) {
-                methods += n + ", ";
-            });
+			// routes.root_route.verbs.forEach(function(n) {
```
remove any old code!

imalberto replied:
Gone !

caridy commented on lib/app/addons/ac/config.common.js:
```diff
@@ -95,7 +95,7 @@ YUI.add('mojito-config-addon', function(Y, NAME) {
          */
         getRoutes: function() {
-            return this._page.routes;
+            return (new Y.mojito.RouteMaker(this._page.routes)).getComputedRoutes();
```
can we cache this? one per request? probably attaching it to `_page` object the first time it is used?

imalberto replied:
Fixed. Adding it to `this._page.computedRoutes`.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,254 +6,299 @@
+    it based on a set of parameters.
 
+    @method resolveParams
+    @protected
```
this is really private I think

imalberto replied:
Fixed.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,254 +6,299 @@
+            // the cache builder
+            params = route.dispatch.params;
+
+        pathParts.forEach(function (pathPart) {
```
this will not work in old browsers, this is a common affinity file. Use `Y.Object.each` instead.

imalberto replied:
Fixed.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -6,254 +6,299 @@
+            name: name,
+            verbs: verbs,
+            params: route.dispatch.params,
+            regex: {},
```
why not using regex as regex? who is using this structure aside from us?

imalberto replied:
In Mojito today, the semantic and syntax for `regex` is different from `ext_match`.

I think once we remove the transformation later in a not-so-distant future, this won't be needed anymore. For now, might be better to keep the route configuration format the same as what it is today, for BC sake.

imalberto commented:
@caridy Added some comments.

caridy commented:
Yeah, I still need more time to finish the review.

imalberto committed:
Updated from feedback.

imalberto committed:
All unit tests passing.

imalberto committed:
Linted. This was causing CI build breakage.

imalberto committed:
Added HISTORY_next.md to track changes on the #next branch

isao commented on lib/dispatcher.js:
```diff
@@ -18,6 +18,7 @@ Given a `command`, dispatches it to the Mojito engine.
 
 var liburl = require('url'),
     debug = require('debug')('mojito:dispatcher'),
+    methods = require('methods'),
```
this module isn't listed as a dependency in package.json. Did you mean this?

```
methods = require('express/node_modules/methods'),
```

imalberto replied:
Its added here: https://github.com/yahoo/mojito/pull/1191/files#L10R24

isao replied:
thx, my mistake.. I had started rebasing onto develop, and.. :/

isao commented on lib/dispatcher.js:
```diff
@@ -56,6 +66,34 @@ module.exports = {
+            if (!mojito.routes) {
+                debug('building up route cache');
+                mojito.routes = {};
+                methods.forEach(function (method) {
```
The package "methods" exports an array of HTTP methods, including a bunch of esoteric ones. But, unless I'm missing something, it doesn't seem to be needed-- this line iterates over the methods array, but is only actionable if it matches a key in `req.app.routes`.

Would it be better to just iterate over `Object.keys(req.app.routes)`? That'd generally be just `['get', 'post']` for web apps

caridy replied:
Yes, @isao is right, you should write this routine by iterating over the methods defined by `Object.keys(req.app.routes)`, not need to use `methods`.

imalberto replied:
Thanks ! Adding diff here for reference: https://github.com/imalberto/mojito/commit/74fe35abf88af0155871a7c0d6eac37c8fec8276

caridy commented:
ok, we can take this for a spin, and we can re-iterate on top of this.

+1 after fixing the stuff that @isao mentioned.

imalberto committed:
- Removed `methods` from dependencies in `package.json`
- Using `Object.keys(req.app.routes)` instead of `methods.forEach` based on feedback
- Added unit tests
